jobs/microk8s/utils.py

```diff
@@ -0,0 +1,213 @@
+    Raises ValueError if string is not correctly formated """
+    if a == b:
+        return 0
+
+    aparts = a.split('-')
+    bparts = b.split('-')
+
+    # Comparing 0[-X] to 1[-Y]
+    if aparts[0] != bparts[0]:
+        arelease = int(aparts[0])
+        brelease = int(bparts[0])
+        if arelease > brelease:
+            return 1
+        if arelease < brelease:
+            return -1
+    else: # We are now at 1[-X] == 1[-Y]
+        # Comparing 14 to 14-{whatever}
+        if len(aparts) < len(bparts):
+            return 1
+        # Comparing 14-{whatever} to 14
+        if len(aparts) > len(bparts):
+            return -1
+
+        # We are now at 1-{something} == 1-{somethingelse}
+        # Comparing 14-{alpha} to 14-{beta,rc}
+        if "alpha" in aparts[1] and any(r in bparts[1] for r in ['beta', 'rc']):
+            return -1
+        # Comparing 14-{alpha,beta} to 14-rc
+        if any(r in aparts[1] for r in ['beta', 'alpha']) and "rc" in bparts[1]:
+            return -1
+
+        # We are now at 1-{something} == 1-{somethingelse}
+        # were something > somethingelse
+        return 1
+
+
+def compare_releases(a, b):
+    """Compares two version string.
+
+    Returns: 1 if a > b, 0 if a==b, -1 if a < b
+    Raises ValueError if string is not correctly formatted """
+
+    major = 0
+    minor = 1
+    patch = 2
+    revision = 3
+
+    a = a.strip()
+    b = b.strip()
+    if a.startswith('v'):
+        a = a[1:]
+    if b.startswith('v'):
+        b = b[1:]
+
+    if a == b:
+        return 0
+
+    aparts = a.split('.')
+    bparts = b.split('.')
+
+    # Major part of a and b have to be digits
+    amajor = int(aparts[major])
+    bmajor = int(bparts[major])
+
+    # Comparing 1.x to 2.y
+    if amajor != bmajor:
+        if amajor > bmajor:
+            return 1
+        elif amajor < bmajor:
+            return -1
+
+    # Since we reached this spot we have major sections equal
+
+    # Minor part of a and b have to be digits
+    aminor = int(aparts[minor])
+    bminor = int(bparts[minor])
+
+    # Comparing X.1 to X.12
+    if aminor != bminor:
+        if aminor > bminor:
+            return 1
+        elif aminor < bminor:
+            return -1
+
+    # Since we reached this spot we have major and minor sections equal
+    patch_compare = compare_patch_with_alpha(aparts[patch], bparts[patch])
+    if patch_compare != 0:
+        return patch_compare
+
+    # Since we reached this spot we have major, minor and patch sections equal
+    # We know revision numbers exist because we have unequal version strings
+    arevision = int(aparts[revision])
+    brevision = int(bparts[revision])
+    if arevision != brevision:
+        if arevision > brevision:
+            return 1
+        elif arevision < brevision:
+            return -1
+
+    # Since we reached this spot we have major, minor, patch and revision sections equal
+    # This should have been caught at the very beginning.
+    assert False
+
```

Comment:
They use semver, you could use the semver library to handle this, ie https://github.com/charmed-kubernetes/jenkins/blob/master/jobs/build-snaps/snaps-source.py#L46-L53

Reply:
Awesome, less code less bugs :)